test: migrate animations/dsl to run under bazel

IgorMinar opened the pull request.

IgorMinar committed:
test: migrate animations/dsl to run under bazel

IgorMinar commented on packages/animations/docs/examples/dsl/animation_example.spec.ts:
```diff
@@ -7,7 +7,7 @@
 
 import {$, ExpectedConditions, browser, by, element} from 'protractor';
-import {verifyNoBrowserErrors} from '../../../../../_common/e2e_util';
+import {verifyNoBrowserErrors} from '../../../../private/testing/e2e';
```
I couldn't use @angular/private/testing/e2e path because node doesn't resolve that module id

IgorMinar commented on packages/animations/docs/examples/dsl/index.html:
```diff
@@ -0,0 +1,13 @@
+<html>
+  <head>
+    <meta charset="utf-8">
+    <link rel="shortcut icon"type="image/x-icon" href="data:image/x-icon;,">
```
this is a neat trick to prevent 404s - we should use it in all the servers that autogenerate index.html

skreborn replied:
This lines is missing a space between the attributes `rel` and `type`.

IgorMinar replied:
thanks, but I removed this file in the latter commits.

IgorMinar commented on packages/animations/docs/examples/dsl/index.html:
```diff
@@ -0,0 +1,13 @@
+  </head>
+  <body>
+    <example-app></example-app>
+    <!-- <script src="ngdeps/node_modules/zone.js/dist/zone.min.js"></script> -->
```
I tried to load zone.js in multiple ways.. all were ugly, so I ended up bundling it into the main bundle.

IgorMinar commented on packages/animations/docs/examples/dsl/tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+{
```
this tsconfig is just for IDE support..

IgorMinar replied:
but I have easy way to know if it's even remotely close to what bazel is using.

alexeagle replied:
Sorry, is that a question? You can see bazels tsconfig in the bin dir

mary-poppins commented:
You can preview 05dad26 at https://pr27390-05dad26.ngbuilds.io/.

mary-poppins commented:
You can preview a384937 at https://pr27390-a384937.ngbuilds.io/.

IgorMinar commented on packages/animations/docs/examples/dsl/BUILD.bazel:
```diff
@@ -0,0 +1,59 @@
+    bootstrap = [
+        "@ngdeps//node_modules/zone.js:dist/zone.js"
+    ],
+    entry_module = "main",
```
- this doesn't boostrap the app (the entry module factory is loaded, but not evaluated)
- ergonomics issue: mistyping the entry module name doesn't result in error, instead the error is silently ignored (see https://github.com/bazelbuild/rules_typescript/issues/336)

alexeagle replied:
check the docs - The entrymodule should be the AMD module name of the entry module such as `"_main/src/index"ts_devserver concats the following snippet after the bundle to load the application:require(["entry_module"]);`

If you look inside your bundle you'll see the module you're trying to load is `angular/packages/animations/docs/examples/dsl/main`

The right answer here is to take a label instead of a string.

IgorMinar replied:
but `_main/src/`is something that comes from the rule and not from the user code - all the apis should always be configured from the user/source code perspective and not from whatever the rule is doing under the hood.

mary-poppins commented:
You can preview 0f01c81 at https://pr27390-0f01c81.ngbuilds.io/.

IgorMinar commented on packages/animations/docs/examples/dsl/dsl_example.spec.ts:
```diff
@@ -7,7 +8,7 @@
 
 import {$, ExpectedConditions, browser, by, element} from 'protractor';
-import {verifyNoBrowserErrors} from '../../../../../_common/e2e_util';
+import {verifyNoBrowserErrors} from '../../../../private/testing/e2e';
```
unlike in browser code, I can't get this node module to resolve if set to `@angular/private/testing/e2e`.

IgorMinar commented on packages/animations/docs/examples/dsl/BUILD.bazel:
```diff
@@ -0,0 +1,59 @@
+    server = ":server",
+    deps = [
+        ":e2e_lib",
+        "//packages/bazel/src/protractor/utils",
```
this dep is here because of the `:on_prepare` script - could we make the two deps "travel" together?

I should be able to specify that "//packages/private/testing/e2e:on_prepare" depends on ""//packages/bazel/src/protractor/utils" at runtime, so if anyone uses the former, the later should come along with it. Can this be done?

mary-poppins commented:
You can preview a5900eb at https://pr27390-a5900eb.ngbuilds.io/.

IgorMinar commented on packages/animations/docs/examples/dsl/BUILD.bazel:
```diff
@@ -0,0 +1,59 @@
+    deps = [
+        ":e2e_lib",
+        "//packages/bazel/src/protractor/utils",
+        "//packages/private/testing/e2e",
```
similar to the previous dep: this dep is here because `:e2e_lib` needs it at runtime. I already had to provide         "//packages/private/testing/e2e" as a build-time dep and unless that dep is purely comprised of d.ts files, any .js files should come along with it to wherever e2e_lib's js files are used.

why don't the dependencies work like this? is it that we haven't set up the rules in this way, or is there something fundamental in bazel that prevents this?

IgorMinar commented on packages/animations/docs/examples/dsl/BUILD.bazel:
```diff
@@ -0,0 +1,59 @@
+        "//packages/bazel/src/protractor/utils",
+        "//packages/private/testing/e2e",
+        "@ngdeps//protractor",
+        "@ngdeps//selenium-webdriver",
```
these last two deps are "peerDeps" of the protractor_web_test_suite rule - I think we should make these an explicit attribute similar to how a compiler is provided to other rules.

IgorMinar commented on packages/animations/docs/examples/dsl/main.ts:
```diff
@@ -0,0 +1,8 @@
+import {platformBrowser} from '@angular/platform-browser';
+import {AppModuleNgFactory} from './dsl_example.ngfactory';
```
this import from nowhere will go away with ivy

IgorMinar commented on packages/private/testing/e2e/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+    srcs = ["index.ts"],
+)
+
+filegroup(
```
I was surprised to find out that there is no way to specify a js_library. Is that by design or we just didn't get to it?

the biggest downside of using `filegroup` here is that I can't specify runtime dependencies of this js file right here.

IgorMinar commented on packages/private/testing/e2e/index.ts:
```diff
@@ -0,0 +1,26 @@
+declare var browser: any;
+declare var expect: any;
+
+// TODO (juliemr): remove this method once this becomes a protractor plugin
```
this should really be part of protractor. @vikerman is there a reason why we wouldn't want it there?

IgorMinar commented on packages/private/testing/e2e/protractor.on_prepare.js:
```diff
@@ -0,0 +1,22 @@
+const protractorUtils = require('@angular/bazel/protractor-utils');
+const protractor = require('protractor');
+
+module.exports = async function(config) {
```
this config seems to be required for all of our protractor tests - shouldn't we move this code to @angular/bazel ?

mary-poppins commented:
You can preview af9bce4 at https://pr27390-af9bce4.ngbuilds.io/.

mary-poppins commented:
You can preview da44a5b at https://pr27390-da44a5b.ngbuilds.io/.

mary-poppins commented:
You can preview 68990f3 at https://pr27390-68990f3.ngbuilds.io/.

mary-poppins commented:
You can preview 364e482 at https://pr27390-364e482.ngbuilds.io/.

mary-poppins commented:
You can preview 6a6e97d at https://pr27390-6a6e97d.ngbuilds.io/.

IgorMinar commented:
Yeah but that config can easily get out of sync with this one without me
realizing that.

On Thu, Dec 6, 2018, 7:17 AM Alex Eagle <notifications@github.com wrote:

> *@alexeagle* commented on this pull request.
> ------------------------------
>
> In packages/animations/docs/examples/dsl/tsconfig.json
> <https://github.com/angular/angular/pull/27390#discussion_r239490923>:
>
> > @@ -0,0 +1,23 @@
> +{
>
> Sorry, is that a question? You can see bazels tsconfig in the bin dir
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/27390#discussion_r239490923>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6Ojv3Y_cLs8jMA2SfgHeRK1FXqaXks5u2TT8gaJpZM4Y80W1>
> .
>

mary-poppins commented:
You can preview 73fb5ab at https://pr27390-73fb5ab.ngbuilds.io/.

mary-poppins commented:
You can preview a3088c2 at https://pr27390-a3088c2.ngbuilds.io/.

IgorMinar committed:
build: add a macro for protractor_web_test_suite that sets the default values

we don't wan't to have too much boilerplate for this rule all over angular/angular, so this macro
sets the most common defaults.

IgorMinar committed:
feat(bazel): protractor_web_test_suite now supports file .e2e.ts extension

I refactored an existing test rather than creating a new one because we have two different
tests and one of them still uses .test.ts extension. So adding a new test is not necessary.

mary-poppins commented:
You can preview 82c2fc4 at https://pr27390-82c2fc4.ngbuilds.io/.

mary-poppins commented:
You can preview d32a874 at https://pr27390-d32a874.ngbuilds.io/.

mary-poppins commented:
You can preview 4ab13a6 at https://pr27390-4ab13a6.ngbuilds.io/.

mary-poppins commented:
You can preview 2a30320 at https://pr27390-2a30320.ngbuilds.io/.

mary-poppins commented:
You can preview 9193976 at https://pr27390-9193976.ngbuilds.io/.

IgorMinar committed:
test: print a human-readable error message when e2e test fails due to client-side errors/warnings

Before:

```
Failures:
1) animation example index view should list out the current collection of items
  Message:
    Expected $.length = 1 to equal 0.
    Expected $[0] = true to equal undefined.
```

After:

```
Failures:
1) animation example index view should list out the current collection of items
  Message:
    Failed: Expected no warnings or errors in the browser console, but found:

    http://localhost:37605/dsl/ - Failed to load resource: the server responded with a status of 404 (Not Found)

```

mary-poppins commented:
You can preview 1f41a54 at https://pr27390-1f41a54.ngbuilds.io/.

IgorMinar committed:
ci: exclude packages/**/docs/** from bot's g3 impact check

These files shouldn't be synced into google3.

PS: we need to make this change in copybara config as well in g3.

mary-poppins commented:
You can preview 00edecc at https://pr27390-00edecc.ngbuilds.io/.

mary-poppins commented:
You can preview 49e6824 at https://pr27390-49e6824.ngbuilds.io/.

IgorMinar commented:
obsolete
